hal/src/samd51/qspi.rs

```diff
@@ -0,0 +1,219 @@
+        mclk: &mut MCLK,
+        port: &mut Port,
+        qspi: QSPI,
+        sck: Pb10<Input<Floating>>,
+        cs:  Pb11<Input<Floating>>,
+        io0: Pa8<Input<Floating>>,
+        io1: Pa9<Input<Floating>>,
+        io2: Pa10<Input<Floating>>,
+        io3: Pa11<Input<Floating>>,
+    ) -> Qspi {
+        let sck = sck.into_function_h(port);
+        let cs = cs.into_function_h(port);
+        let io0 = io0.into_function_h(port); 
+        let io1 = io1.into_function_h(port); 
+        let io2 = io2.into_function_h(port); 
+        let io3 = io3.into_function_h(port); 
+
+        mclk.apbcmask.modify(|_, w| w.qspi_().set_bit());
+        mclk.ahbmask.modify(|_, w| w.qspi_().set_bit());
+        mclk.ahbmask.modify(|_, w| w.qspi_2x_().clear_bit());
+
+        qspi.ctrla.modify(|_, w| w.swrst().set_bit());
+        qspi.baud.modify(|_, w| {
+            unsafe { w.baud().bits(14) }; //120MHz / (14+1) = 8MHz
+            
+            // SPI MODE 0
+            w.cpol().clear_bit();
+            w.cpha().clear_bit()
+        });
+
+        qspi.ctrlb.modify(|_, w| {
+            w.mode().memory();
+            w.csmode().noreload();
+            w.csmode().lastxfer();
+            w.datalen()._8bits()
+        });
+
+        qspi.ctrla.modify(|_, w| w.enable().set_bit());
+
+        Qspi { 
+            qspi,
+            sck,
+            cs,
+            io0,
+            io1,
+            io2,
+            io3,
+        }
+    }
+    
+    unsafe fn run_write_instruction(
+        &self, 
+        command: Command, 
+        addr: u32, 
+        buf: &[u8]
+    ) {
+        if command == Command::EraseSector || command == Command::EraseBlock {
+            self.qspi.instraddr.write(|w| { w.addr().bits(addr) });
+        }
+        self.qspi.instrctrl.write(|w| w.instr().bits(command.bits()));
+        let _ = self.qspi.instrframe.read().bits();
+        if buf.len() > 0 {
+            core::ptr::copy(buf.as_ptr(), (QSPI_AHB + addr) as *mut u8, buf.len()); 
+        }
+
+        self.qspi.ctrla.write(|w| {
+            w.enable().set_bit();
+            w.lastxfer().set_bit()
+        });
+
+        while self.qspi.intflag.read().instrend().bit_is_clear() {}
+        self.qspi.intflag.modify(|_, w| w.instrend().set_bit());
+    }
+
+    unsafe fn run_read_instruction(
+        &self, 
+        command: Command, 
+        addr: u32, 
+        buf: &mut [u8]
+    ) {
+        self.qspi.instrctrl.write(|w| w.instr().bits(command.bits()));
+        let _ = self.qspi.instrframe.read().bits();
+        if buf.len() > 0 {
+            core::ptr::copy((QSPI_AHB + addr) as *mut u8, buf.as_mut_ptr(), buf.len());  
+        }
+
+        self.qspi.ctrla.write(|w| {
+            w.enable().set_bit();
+            w.lastxfer().set_bit()
+        });
+
+        while self.qspi.intflag.read().instrend().bit_is_clear() {}
+        self.qspi.intflag.modify(|_, w| w.instrend().set_bit());
+    }
+
+    pub fn run_command(&self, command: Command) {
```

Comment:
Would it make sense for these to return a result?

Could we add a docstring for these public methods?